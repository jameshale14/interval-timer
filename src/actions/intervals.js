import database from '../firebase/firebase'

//GET_INTERVALS
export const getIntervals = () => ({
  type: 'GET_INTERVALS',
})

//START_GET_INTERVALS

//CREATE_INTERVAL
export const createInterval = (interval) => ({
  type: 'CREATE_INTERVAL',
  interval
})

//START_CREATE_INTERVAL

//UPDATE_INTERVAL
export const updateInterval = (id, updates) => ({
  type: 'UPDATE_INTERVAL',
  id,
  updates
})

//START_UPDATE_INTERVAL

//DELETE_INTERVAL
export const deleteInterval = (id) => ({
  type: 'DELETE_INTERVAL',
  id
})

//START_DELETE_INTERVAL

export const startDeleteInterval = (id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/intervals/${id}`).remove()
      .then(() => {
        dispatch(deleteInterval(id))
      })
  }
}

//SET_INTERVALS
export const setIntervals = (intervals) => ({
  type: 'SET_INTERVALS',
  intervals
})

//START_SET_INTERVALS
export const startSetIntervals = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/intervals`).once('value')
      .then((snapshot) => {
        const intervals = []
        snapshot.forEach((childSnapshot) => {
          intervals.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setIntervals(intervals))
      })
  }
}