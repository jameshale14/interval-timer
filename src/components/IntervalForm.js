import React from 'react'
import PropTypes from 'prop-types'

export default class IntervalForm extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      stepError: undefined,
      saveError: undefined,
      name: props.interval ? props.interval.name : '',
      steps: props.interval ? props.interval.steps : [],
      newStep: {
        type: 'Activity',
        name: '',
        duration: ''
      }
    }
  }

  handleAddStep = (e) => {
    e.preventDefault()
    if (this.state.newStep.type == '' || this.state.newStep.name == '' || this.state.newStep.duration == 0) {
      this.setState(() => ({ stepError: 'Please fill in all fields' }))
    } else {
      this.setState((prevState) => {
        const newStep = { ...this.state.newStep }
        return {
          steps: [...prevState.steps, newStep],
          stepError: undefined
        }
      })
      // this.setState(() => ({  }))
    }
  }

  onTypeChange = (e) => {
    const type = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.type = type
      return { newStep }
    })
  }

  onNameChange = (e) => {
    const name = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.name = name
      return { newStep }
    })
  }

  onIntervalNameChange = (e) => {
    const name = e.target.value
    this.setState(() => ({ name }))
  }

  onDurationChange = (e) => {
    const duration = e.target.value
    this.setState((prevState) => {
      const newStep = prevState.newStep
      newStep.duration = duration
      return { newStep }
    })
  }

  handleOnSave = (e) => {
    e.preventDefault()

    if (!this.state.name || this.state.steps.length === 0) {
      this.setState(() => ({ saveError: 'Please provide a routine name and interval steps.' }))
    } else {
      this.setState(() => ({ saveError: undefined }))

      const interval = {
        name: this.state.name,
        steps: this.state.steps
      }

      this.props.onSubmit(interval)
    }

  }

  handleRemoveStep = (e) => {
    e.preventDefault()
    const stepIndex = parseInt(e.currentTarget.value)
    this.setState((prevState) => {
      const steps = prevState.steps.filter((step, index) => {
        return index != stepIndex
      })

      return ({
        steps
      })

    })
  }

  render() {
    return (
      <div>
        <button id='saveInterval' onClick={this.handleOnSave}>Save</button>
        {this.state.saveError && <p>{this.state.saveError}</p>}
        <input
          type='text'
          placeholder='name'
          value={this.state.name}
          onChange={this.onIntervalNameChange}
        />
        {
          this.state.steps.map((step, stepIndex) => {
            return (
              <div key={stepIndex} >
                <p>Type: {step.type}</p>
                <p>Name: {step.name}</p>
                <p>Duration: {step.duration}</p>
                <button onClick={this.handleRemoveStep} value={stepIndex}>Remove</button>
              </div>
            )
          })
        }
        <div>
          {this.state.stepError && <p>{this.state.stepError}</p>}
          <form onSubmit={this.handleAddStep}>
            <select
              value={this.state.newStep.type}
              onChange={this.onTypeChange}
              onBlur={this.onTypeChange}
            >
              <option value='Activity'>Activity</option>
              <option value='Rest'>Rest</option>
            </select>

            <input
              type='text'
              placeholder='Step name'
              value={this.state.newStep.name}
              onChange={this.onNameChange}
            />
            <input
              type='number'
              placeholder='Duration (seconds)'
              step='1'
              value={this.state.newStep.duration}
              onChange={this.onDurationChange}
            />
            <button onClick={this.handleAddStep}>Add Step</button>
          </form>
        </div>
      </div>
    )
  }
}

IntervalForm.propTypes = {
  interval: PropTypes.object,
  onSubmit: PropTypes.func
}
