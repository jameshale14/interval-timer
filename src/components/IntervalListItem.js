import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

const ExpenseListItem = ({ id, name }) => (

  <Link to={`/edit/${id}`}>
    <>
      <h3>{name}</h3>
    </>
  </Link>

)

ExpenseListItem.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string
}

export default ExpenseListItem