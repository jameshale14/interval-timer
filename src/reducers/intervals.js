export default (state = {}, action) => {
  switch (action.type) {
    case 'GET_INTERVALS':
      return state
    case 'CREATE_INTERVAL':
      return [...state, action.interval]
    case 'UPDATE_INTERVAL':
      return state.map((interval) => {
        if (interval.id === action.id) {
          return {
            ...interval,
            ...action.updates
          }
        } else {
          return interval
        }
      })
    case 'DELETE_INTERVAL':
      return state.filter((interval) => {
        return interval.id !== action.id
      })
    case 'SET_INTERVALS':
      return action.intervals
    default:
      return state
  }
}