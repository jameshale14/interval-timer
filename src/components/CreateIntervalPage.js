import React from 'react'
import { connect } from 'react-redux'
import IntervalForm from './IntervalForm'

export class CreateIntervalPage extends React.Component {
  onSubmit = (interval) => {
    this.props.startCreateInterval(interval)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <IntervalForm onSubmit={this.onSubmit} />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startCreateInterval: (interval) => dispatch(startCreateInterval(interval))
})

export default connect(undefined, mapDispatchToProps)(CreateIntervalPage)