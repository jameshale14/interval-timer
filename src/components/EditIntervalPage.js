import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import IntervalForm from './IntervalForm'
import { startUpdateInterval } from '../actions/intervals'

export class EditIntervalPage extends React.Component {

  onSubmit = (interval) => {
    this.props.startUpdateInterval(this.props.interval.id, interval)
    this.props.history.push('/')
  }

  render() {
    return (
      <>
        <h3>Edit Interval</h3>
        <div>
          <IntervalForm onSubmit={this.onSubmit} interval={this.props.interval} />
        </div>
      </>
    )
  }
}

EditIntervalPage.propTypes = {
  interval: PropTypes.object,
  startUpdateInterval: PropTypes.func,
  history: PropTypes.object
}

const mapStateToProps = (state, props) => ({
  interval: state.intervals.find((interval) => interval.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
  startUpdateInterval: (id, interval) => dispatch(startUpdateInterval(id, interval))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditIntervalPage)