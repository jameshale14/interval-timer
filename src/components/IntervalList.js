import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import IntervalListItem from './IntervalListItem'

export const IntervalList = (props) => (
  <div>
    <h1>Intervals</h1>

    {(props.intervals && props.intervals.length) > 0 ? (props.intervals.map((interval) => {
      return (
        <IntervalListItem key={interval.id} {...interval} />
      )
    }))
      : (
        <p>No intervals</p>
      )
    }
  </div>
)

IntervalList.propTypes = {
  intervals: PropTypes.array
}

const mapStateToProps = (state) => ({
  intervals: state.intervals
})

export default connect(mapStateToProps)(IntervalList)