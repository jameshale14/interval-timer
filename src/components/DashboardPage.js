import React from 'react'
import { Link } from 'react-router-dom'
import IntervalList from './IntervalList'

const DashBoardPage = () => (
  <div>
    <Link className='button' to='/create'>New Interval</Link>
    <IntervalList />
  </div>
)

export default DashBoardPage 