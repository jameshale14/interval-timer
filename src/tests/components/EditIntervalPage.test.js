import React from 'react'
import { shallow } from 'enzyme'
import { EditIntervalPage } from '../../components/EditIntervalPage'
import { intervals } from '../fixtures/intervals'

let startUpdateInterval, history, wrapper

beforeEach(() => {
  startUpdateInterval = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<EditIntervalPage startUpdateInterval={startUpdateInterval} history={history} interval={intervals[1]} />)
})

test('should render EditIntervalPage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle onSubmit', () => {
  wrapper.find('IntervalForm').prop('onSubmit')(intervals[1])
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(startUpdateInterval).toHaveBeenLastCalledWith(intervals[1].id, intervals[1])
})