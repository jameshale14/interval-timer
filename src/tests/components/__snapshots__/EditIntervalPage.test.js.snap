// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditIntervalPage correctly 1`] = `
<Fragment>
  <h3>
    Edit Interval
  </h3>
  <div>
    <IntervalForm
      interval={
        Object {
          "id": "2",
          "name": "meditation focus",
          "steps": Array [
            Object {
              "duration": 30000,
              "name": "focus",
              "type": "activity",
            },
            Object {
              "duration": 15000,
              "name": "rest",
              "type": "rest",
            },
            Object {
              "duration": 30000,
              "name": "focus",
              "type": "activity",
            },
            Object {
              "duration": 15000,
              "name": "rest",
              "type": "rest",
            },
          ],
        }
      }
      onSubmit={[Function]}
    />
  </div>
</Fragment>
`;
